6 add archetype functions

crispy-wonton opened the pull request:
Addresses #6 

# Description

Summary of changes:
- add archetypes functions to classify properties into 8 given archetypes
- add functions to generate cost quantiles and create a clean final output dataframe of cost percentiles by property archetype
- update main() and run() functions to use the new functions listed above and produce basic target output of dataframe and excel file (saved to S3) of adjusted cost percentiles by property archetype

NB I will add a descriptive README in a separate PR.

# Instructions for Reviewer

Please could you:
- general check for errors and issues
- test the code runs for you with any flags you want from the args. This is the line that I used to produce a test file that's saved to S3 `python asf_heat_pump_affordability/pipeline/produce_costs_dataset.py --mcs_epc_join_date 231009 --cost_year_min 2021 --cpi_data_year 2023 --cost_quantiles 0 0.25 0.5 0.75 1 --save_to_s3`
- leave off `--save_to_s3` flag if not wanting to save to s3
- you can also test by importing `main()` as before. I have also checked this works for me.
- please let me know if you have any ideas to further streamline the `classify_dict_archetypes_masks` function. I've left a comment on the file to explain my choices.
- let me know if you think the functions in `generate_cost_percentiles.py` are too prescriptive

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
add property archetypes classification functions

crispy-wonton committed:
add functions to generate cost percentiles and convert to dataframe

crispy-wonton committed:
add archetypes and cost percentiles functions into run script
run script generates basic target output: dataframe with cost percentiles by property archetype

crispy-wonton commented on asf_heat_pump_affordability/pipeline/archetypes.py:
```diff
@@ -0,0 +1,102 @@
+from asf_heat_pump_affordability import config
+
+
+def classify_dict_archetypes_masks(
```
Here I tried to give the function some flexibility - i.e. ages_mapping can be updated if desired - but hard coded the property types rather than overcomplicate the code. If archetypes change, the code will need to be updated or a new function added.

The alternative option I could think of was hard coding helper functions for each archetype with year split and then having one `classify_dict_archetypes_masks()` function which generates the dict using only helper functions and no hard coding.

Ultimately I went with the choice here in the hopes that it is slightly more flexible and would reduce amount of code changes required if archetypes change slightly.

danlewis85 replied:
I'm happy with this as is. for me the important thing is that you could in theory supply any mapping of archetypes-boolean filter and generate a result.

crispy-wonton committed:
correct quantile function names and improve doc strings

crispy-wonton committed:
update main function doc string and fix argparse bug

danlewis85 commented on asf_heat_pump_affordability/pipeline/produce_costs_dataset.py:
```diff
@@ -114,20 +114,20 @@ def main(
     archetypes_dict = archetypes.classify_dict_archetypes_masks(mcs_epc_inf)
 
-    # Get cost percentiles
-    if not cost_percentiles:
```
NB This works if `cost_percentiles` is a list, because `bool([1,2,3])` evaluates to `True`, however it doesn't work if `cost_percentiles` is a numpy array (which you permit in the type hint), so `bool(numpy.array([1,2,3]))` returns a ValueError (The truth value of an array with more than one element is ambiguous. Use a.any() or a.all())

As such in this case, you most likely want to evaluate as: `if cost_percentiles is not None:`

crispy-wonton replied:
Ah great catch thank you! I didn't know.

danlewis85 commented on asf_heat_pump_affordability/pipeline/generate_cost_percentiles.py:
```diff
@@ -0,0 +1,91 @@
+def generate_dict_cost_percentiles_by_archetype(
+    cost_series: pd.Series,
+    archetypes_masks: dict,
+    percentiles: list or np.array,
```
This could be more [generic](https://mypy.readthedocs.io/en/stable/builtin_types.html#generic-types), e.g.

`from typing import Iterable`

`percentiles: Iterable[float]`

crispy-wonton replied:
Oh yeah thanks!

danlewis85 commented on asf_heat_pump_affordability/pipeline/generate_cost_percentiles.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def generate_df_cost_percentiles_by_archetype_formatted(
+    archetype_costs_dict: dict, percentiles: list or np.array, ref_year: int
```
As above.

danlewis85 reviewed (changes requested):
Fix the issue with the `if not cost_quantiles:` conditional in produce_costs_dataset.py and update some of the type hinting and this should be fine to merge.

crispy-wonton committed:
apply changes requested in PR review
- improve iterable type hinting
- fix bug to allow array input for cost_quantiles param in main() function
